### PR TITLE
[otel-arrow/receiver] Receiver concurrency fixes; readability improvements & restructuring

### LINKED ISSUE #204: Receiver crash report: concurrency bug
The v0.23.0 receiver includes a potential crash due to a concurrency violation since #181.

The Arrow consumer needs to be used sequentially from a single goroutine. I believe this was the intention, and the code needs to be refactored to improve readability, since this bug was not uncovered during review.

### PR DESCRIPTION
**Pull request by jmacd**
Restructure receiver code to improve readability. There are a number of metrics that are incremented when a batch starts being processed and are decremented when the batch is finished, but the control flow that maintained the balance of these updates was convoluted.

The root-cause of #204 is that Arrow batches meant for a consumer to be processed in order were processed out-of-order. There was a large function body which served two purposes: consume Arrow data of the appropriate kind, enter data for the pipeline to consume next. This had to be split into two parts and should have been done as part of #181. (I, as reviewer, missed this and find, in hindsight, that the code is not easy to follow.)

This improves the code structure by moving all stateful aspects of starting/finishing a request into a new `inFlightData` object which has a deferrable method to finish the request. Here, we keep:

1. The `inFlightWG` done count
2. The active requests metric
3. The active items metric
4. The active bytes metric
5. The bytes-acquired from the semaphore
6. A per-request span covering Arrow decode
7. Netstat-related instrumentation

Authorization now happens before acquiring from the semaphore. 

A number of `fmt.Errorf()` calls are replaced with `status.Errorf(...)` and a specific error code. The tests are updated to be more specific.  Several Arrow tests were accidentally canceling the test before an expected error condition was actually tested, they have been audited and improved.

One new concurrent-receiver test was added.

Fixes #204.

### CODE REVIEW

File: collector/receiver/otelarrowreceiver/internal/arrow/arrow.go [hunk 1]
```diff
@@ -552,235 +730,177 @@ func (r *Receiver) flushSender(serverStream anyStreamServer, pendingCh <-chan ba
...
-	// not necessarily break the stream.
-	if authErr != nil {
-		err = authErr
-	} else {
-		err = r.processRecords(ctx, method, arrowConsumer, req, response, sizeHeaderFound)
-	}
-
-	return err
-}
+// consumeBatch applies the batch to the Arrow Consumer, returns a
```
Comment (jmacd) [inline collector/receiver/otelarrowreceiver/internal/arrow/arrow.go:747]: Reviewer notes: The function body in the old code was once large and repetitive. The key bugfix here is to split this function into two parts, one that runs sequentially to consume Arrow data, and one that runs concurrently for the pipeline to process data.

There is less repetition now, due to the consolidation into `inFlightData`.

**Comment by jmacd** [comment]
@kristinapathak Proposing to add you as a reviewer in this repository. This change is a substantial rewrite but not a substantial change of behavior, so it's worth a careful review and should help you learn the receiver code very well. Thank you! 😁 🚀

File: collector/receiver/otelarrowreceiver/internal/arrow/arrow.go [hunk 2]
```diff
@@ -370,163 +369,338 @@ func (r *Receiver) anyStream(serverStream anyStreamServer, method string) (retEr
...
-		return fmt.Errorf("breaking stream: %w", err)
+	// Authorize the request, if configured, prior to acquiring resources.
+	if r.authServer != nil {
+		var authErr error
+		inflightCtx, authErr = r.authServer.Authenticate(inflightCtx, authHdrs)
+		if authErr != nil {
+			flight.replyToCaller(status.Error(codes.Unauthenticated, authErr.Error()))
+			return nil
+		}
 	}
```
Comment (kristinapathak) [inline collector/receiver/otelarrowreceiver/internal/arrow/arrow.go]: Can't we do this before the `prevAcquiredBytes` logic?

Comment (kristinapathak) [reply collector/receiver/otelarrowreceiver/internal/arrow/arrow.go]: I understand wanting to authenticate before doing any work, but some amount of memory has already been used by this point.  Is it worth it to use additional resources to authenticate before memory limiting?

File: collector/receiver/otelarrowreceiver/internal/arrow/arrow.go [hunk 3]
```diff
@@ -552,235 +730,177 @@ func (r *Receiver) flushSender(serverStream anyStreamServer, pendingCh <-chan ba
...
+// if any errors are permanent, returns a permanent error.
+func (r *Receiver) consumeData(ctx context.Context, data any, flight *inFlightData) error {
+	var anyPermanent bool
+	var wholeErr error
+	type unwrapper interface{ Unwrap() error }
+
+	oneOp := func(err error) {
+		if consumererror.IsPermanent(err) {
+			anyPermanent = true
+			err = err.(unwrapper).Unwrap()
```
Comment (kristinapathak) [inline collector/receiver/otelarrowreceiver/internal/arrow/arrow.go]: Why are we unwrapping the error?

**Review by kristinapathak — APPROVED** [review]
Left a few questions.  The arrow.go file is long - might be worth splitting up in a separate PR. 🙂

File: collector/receiver/otelarrowreceiver/internal/arrow/arrow_test.go [hunk 4]
```diff
@@ -140,6 +143,7 @@ func (ctc *commonTestCase) putBatch(payload *arrowpb.BatchArrowRecords, err erro
 func (ctc *commonTestCase) doAndReturnGetBatch(ctx context.Context) func() (*arrowpb.BatchArrowRecords, error) {
 	return func() (*arrowpb.BatchArrowRecords, error) {
 		select {
+		// @@@ WHY IS THIS NOT DONE
```
Comment (lquerel) [inline collector/receiver/otelarrowreceiver/internal/arrow/arrow_test.go]: Is it a ToDo ?

File: collector/receiver/otelarrowreceiver/internal/arrow/arrow.go [hunk 5]
```diff
@@ -370,163 +369,338 @@ func (r *Receiver) anyStream(serverStream anyStreamServer, method string) (retEr
 	}()
 	defer r.recoverErr(&retErr)
 
+	// doneCancel allows an error in the sender/receiver to
+	// interrupt the corresponding thread.
 	doneCtx, doneCancel := context.WithCancel(streamCtx)
 	defer doneCancel()
+
+	// streamErrCh returns up to two errors from the sender and
+	// recevier threads started below.
```
Comment (lquerel) [inline collector/receiver/otelarrowreceiver/internal/arrow/arrow.go]:
```suggestion
	// receiver threads started below.
```

File: collector/receiver/otelarrowreceiver/internal/arrow/arrow.go (hunk 1 above)
Comment (lquerel) [reply collector/receiver/otelarrowreceiver/internal/arrow/arrow.go:747]: Thanks for this in-context explanation :-)

File: collector/receiver/otelarrowreceiver/otelarrow_test.go [hunk 6]
```diff
@@ -565,12 +576,101 @@ func TestGRPCArrowReceiverAuth(t *testing.T) {
...
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	const itemsPerStream = 10
+	const numStreams = 5
+
+	var wg sync.WaitGroup
+	wg.Add(numStreams)
+
+	for j := 0; j < numStreams; j++ {
```
Comment (lquerel) [inline collector/receiver/otelarrowreceiver/otelarrow_test.go:616]: Would this test really have been able to detect the initial out-of-order arrow message problem?

**Review by lquerel — COMMENTED** [review]
Generally, it seems much clearer to me, but I have a small doubt about the test (see my comment regarding the last unit test). It's probably because I haven't understood everything. With a few additional explanations, I would certainly be able to approve this commit.

File: collector/receiver/otelarrowreceiver/internal/arrow/arrow_test.go (hunk 4 above)
Comment (jmacd) [reply collector/receiver/otelarrowreceiver/internal/arrow/arrow_test.go]: This was a problem I had been investigating, related to the test framework and how gRPC streams work. The test uses a mock gRPC client stream, and the receiver under test calls `Recv()` on the mock client. There is a context underlying the mock client that nothing will cancel, until the test explicitly cancels it. The changes in these tests are to cancel them less, and to wait for other conditions to terminate them properly.

File: collector/receiver/otelarrowreceiver/otelarrow_test.go (hunk 6 above)
Comment (jmacd) [reply collector/receiver/otelarrowreceiver/otelarrow_test.go:616]: This test does not. I have another change with a new end-to-end test that did uncover the problem. Thanks!

File: collector/receiver/otelarrowreceiver/internal/arrow/arrow.go (hunk 3 above)
Comment (jmacd) [reply collector/receiver/otelarrowreceiver/internal/arrow/arrow.go]: This was unnecessary -- I had to remind myself that errors.Unwrap() can take two forms, one is `Unwrap() []error` and that is how `IsPermanent()` works on multierr. I've removed the extra cruft.

File: collector/receiver/otelarrowreceiver/internal/arrow/arrow.go (hunk 2 above)
Comment (jmacd) [reply collector/receiver/otelarrowreceiver/internal/arrow/arrow.go]: I think we can do authentication before resource limiting, and yes the call to Auth can happen before prevAcquiredBytes is calculated.

Do you agree with this order? If the Auth fails afterwards, the acquired resource will be released immediately, so it seems to not matter very much.

**Comment by jmacd** [comment]
Reviewers, please take another look.

**Comment by jmacd** [comment]
For the test flake above, https://github.com/open-telemetry/otel-arrow/issues/206

**Review by lquerel — APPROVED** [review]
LGTM

**Review by kristinapathak — APPROVED** [review]

**Pull request merged**